### DISCUSSION TITLE
Support for pgvector HNSW index

### DISCUSSION THREAD

**Discussion by dmurat** [langchain4j/langchain4j#1957] [Q&A]
I'm just trying to orient myself around langchain4j features and APIs. 

While browsing the source, I noticed that PgVectorEmbeddingStore supports only the IVFFlat index. Is the support for the HNSW index planned?

**Comment by odelpozo-denodo** [comment]
Hello,

I’d like to reopen this thread.

While `pgvector` supports IVFFlat indexes, they are ineffective unless the table is pre-populated with data, making them significantly harder to manage compared to HNSW indexes.

Additionally, even if an HNSW index is created on the table's `embedding` column, PostgreSQL currently ignores it due to how search queries are structured. To leverage HNSW indexes effectively, the search query syntax needs to be modified to the following format:

```sql
WITH top_candidates AS (
    SELECT 
        (2 - (embedding::halfvec(<dimensions>) <=> '[-0.111111111,...]'::halfvec(<dimensions>))) / 2 AS score,
        (embedding::halfvec(<dimensions>) <=> '[-0.111111111,...]'::halfvec(<dimensions>)) AS distance,
        embedding_id,
        embedding,
        text,
        metadata
    FROM <table>
    WHERE <filter>
    ORDER BY embedding::halfvec(<dimensions>) <=> '[-0.111111111,...]'::halfvec(<dimensions>)
    LIMIT <maxResults*n>
)
SELECT score, embedding_id, embedding, text, metadata
FROM top_candidates
WHERE distance <= <minScore>
LIMIT <maxResults>;

```

Do you think implementing this logic would be feasible (not only for `pgvector`, but potentially for other vector stores like Oracle as well)?
